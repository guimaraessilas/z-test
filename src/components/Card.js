import React, { useState } from "react";
import {
  View,
  Text,
  StyleSheet,
  Dimensions,
  TouchableOpacity,
  Image,
} from "react-native";
import { numberToBLR } from "../utils/helper";
import { useDispatch } from "react-redux";
import { remove, add } from "../redux/actions";

const Card = (props) => {
  const CardImage = (props) => {
    const [err, setErr] = useState(false);
    return (
      <Image
        source={{
          uri: err
            ? "https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcSYg54LY5wKBL_envlsopC2rnb0oAFqMhLO0A&usqp=CAU"
            : props.src,
        }}
        onError={() => setErr(true)}
        style={styles.imageCard}
      />
    );
  };
  const dispatch = useDispatch();
  const [qtd, setQtd] = useState(0);

  const addItem = (price) => {
    dispatch(add(price));
    setQtd(qtd + 1);
  };

  const removeItem = (price) => {
    if (qtd > 0) {
      dispatch(remove(price));
      setQtd(qtd - 1);
    }
  };

  return (
    <View style={styles.card}>
      <CardImage style={styles.imageCard} src={props.product.images[0].url} />
      <Text style={styles.cardTitle}>{props.product.title}</Text>
      <Text styles={styles.cardPrice}>
        {numberToBLR(props.product.productVariants[0].price)}
      </Text>
      <View style={styles.cardFooter}>
        <TouchableOpacity
          style={styles.button}
          onPress={() => removeItem(props.product.productVariants[0].price)}
        >
          <Text style={styles.buttonText}>-</Text>
        </TouchableOpacity>
        <Text>{qtd}</Text>
        <TouchableOpacity
          style={styles.button}
          onPress={() => addItem(props.product.productVariants[0].price)}
        >
          <Text style={styles.buttonText}>+</Text>
        </TouchableOpacity>
      </View>
    </View>
  );
};

export default Card;

const styles = StyleSheet.create({
  card: {
    flex: 1,
    alignItems: "center",
    backgroundColor: "white",
    marginHorizontal: 2,
    marginVertical: 5,
    height: 250,
    width: Dimensions.get("window").width / 2.1,
    borderRadius: 5,
    padding: 5,
  },
  cardTitle: {
    fontSize: 16,
    fontWeight: "bold",
    textAlign: "center",
    color: "#212121",
  },
  cardPrice: {
    fontSize: 14,
    color: "#757575",
    textAlign: "center",
  },
  cardFooter: {
    flexDirection: "row",
    width: "70%",
    alignItems: "center",
    justifyContent: "space-around",
  },
  button: {
    height: 40,
    width: 40,
    borderRadius: 100,
    backgroundColor: "#757575",
    alignItems: "center",
    justifyContent: "center",
  },
  buttonText: {
    textAlign: "center",
    fontSize: 30,
    color: "white",
  },
  imageCard: {
    flex: 1,
    resizeMode: "contain",
    width: "100%",
    height: 200,
  },
});
