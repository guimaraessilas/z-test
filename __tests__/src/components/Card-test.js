import "react-native";
import React from "react";
import { create, act } from "react-test-renderer";
import { useDispatch } from "react-redux";
import Card from "../../../src/components/Card";
import { TouchableOpacity } from "react-native";

jest.mock("react-redux");
const mockProps = {
  product: {
    images: [
      {
        url: "mock-image",
      },
    ],
    title: "TÃ­tulo",
    productVariants: [{ price: 10.5 }],
  },
};

const setup = () => create(<Card product={mockProps.product} />);

describe("Should test Card Component ", () => {
  beforeAll(() => {
    useDispatch.mockImplementation(() => jest.fn());
  });

  describe("snapshot", () => {
    it("render correctly", () => {
      const tree = setup().toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  it("Click", () => {
    const wrapper = setup();
    act(() => {
      let arr = wrapper.root.findAllByType(TouchableOpacity);
      arr.map((button) => button.props.onPress());
    });

    expect(useDispatch).toHaveBeenCalled();
  });
});
