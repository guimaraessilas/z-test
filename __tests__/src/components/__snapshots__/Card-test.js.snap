// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should test Card Component  snapshot render correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "borderRadius": 5,
      "flex": 1,
      "height": 250,
      "marginHorizontal": 2,
      "marginVertical": 5,
      "padding": 5,
      "width": 357.1428571428571,
    }
  }
>
  <Image
    onError={[Function]}
    source={
      Object {
        "uri": "mock-image",
      }
    }
    style={
      Object {
        "flex": 1,
        "height": 200,
        "resizeMode": "contain",
        "width": "100%",
      }
    }
  />
  <Text
    style={
      Object {
        "color": "#212121",
        "fontSize": 16,
        "fontWeight": "bold",
        "textAlign": "center",
      }
    }
  >
    TÃ­tulo
  </Text>
  <Text
    styles={
      Object {
        "color": "#757575",
        "fontSize": 14,
        "textAlign": "center",
      }
    }
  >
    R$ 10,50
  </Text>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-around",
        "width": "70%",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#757575",
          "borderRadius": 100,
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "width": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 30,
            "textAlign": "center",
          }
        }
      >
        -
      </Text>
    </View>
    <Text>
      0
    </Text>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#757575",
          "borderRadius": 100,
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "width": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 30,
            "textAlign": "center",
          }
        }
      >
        +
      </Text>
    </View>
  </View>
</View>
`;
